cli/src/commands/git/fetch.rs

```diff
@@ -72,13 +84,45 @@ pub fn cmd_git_fetch(
     args: &GitFetchArgs,
 ) -> Result<(), CommandError> {
     let mut workspace_command = command.workspace_helper(ui)?;
-    let remotes = if args.all_remotes {
-        git::get_all_remote_names(workspace_command.repo().store())?
+    let remote_patterns = if args.all_remotes {
+        vec![StringPattern::everything()]
     } else if args.remotes.is_empty() {
         get_default_fetch_remotes(ui, &workspace_command)?
     } else {
         args.remotes.clone()
     };
+
+    let all_remotes = git::get_all_remote_names(workspace_command.repo().store())?;
+
+    let mut remotes = HashSet::new();
+    let mut unmatched_patterns = Vec::new();
+    for pattern in remote_patterns {
+        let matching_remotes = all_remotes
+            .iter()
+            .filter(|r| pattern.matches(r))
+            .collect_vec();
+        if matching_remotes.is_empty() {
+            unmatched_patterns.push(pattern);
+        } else {
+            remotes.extend(matching_remotes);
+        }
+    }
+
+    match &unmatched_patterns[..] {
+        [] => {} // Everything matched, all good
+        [pattern] if pattern.is_exact() => {
+            return Err(user_error(format!("No git remote named '{pattern}'")))
+        }
+        patterns => {
+            return Err(user_error(format!(
+                "No matching git remotes for patterns: {}",
+                patterns.iter().join(", ")
+            )))
+        }
+    }
+
+    let remotes = remotes.into_iter().map(|s| s.as_str()).collect_vec();
```

Comment:
The order is not deterministic (though the stats output is sorted.) I think we can use `all_remotes.retain(|r| remotes.contains(r))` or something.